lib/router/src/cache.rs

```diff
@@ -1,426 +1,367 @@
+impl<K: Clone + Eq + Hash, V> Cache<K, V> {
+    pub fn new(capacity: usize, expires: Duration) -> (Arc<Mutex<Cache<K, V>>>, PurgeCache<K, V>) {
+        let cache = Self {
             capacity,
+            expires,
+            expirations: DelayQueue::with_capacity(capacity),
             vals: IndexMap::default(),
-            max_idle_age,
-            now: (),
-        }
+        };
+        let cache = Arc::new(Mutex::new(cache));
+        let bg_purge = PurgeCache {
+            cache: Arc::downgrade(&cache),
+            interval: Interval::new_interval(expires),
+        };
+
+        (cache, bg_purge)
     }
-}
 
-impl<K: Hash + Eq, V, N: Now> Cache<K, V, N> {
-    /// Accesses a route.
-    ///
-    /// A mutable reference to the route is wrapped in the returned `Access` to
-    /// ensure that the access-time is updated when the reference is released.
-    pub fn access<Q>(&mut self, key: &Q) -> Option<Access<V, N>>
-    where
-        Q: Hash + Equivalent<K>,
-    {
-        let v = self.vals.get_mut(key)?;
-        Some(v.access(&self.now))
+    pub fn access(&mut self, key: &K) -> Option<Access<K, V>> {
+        let node = self.vals.get_mut(key)?;
+        Some(Access {
+            expires: self.expires,
+            expirations: &mut self.expirations,
+            node,
+        })
     }
 
-    /// Ensures that there is capacity to store an additional route.
-    ///
-    /// Returns a handle that may be used to store an ite,. If there is no available
-    /// capacity, idle entries may be evicted to create capacity.
-    ///
-    /// An error is returned if there is no available capacity.
-    pub fn reserve(&mut self) -> Result<Reserve<K, V, N>, CapacityExhausted> {
+    pub fn reserve(&mut self) -> Result<Reserve<K, V>, CapacityExhausted> {
         if self.vals.len() == self.capacity {
-            // Only whole seconds are used to determine whether a node should be retained.
-            // This is intended to prevent the need for repetitive reservations when
-            // entries are clustered in tight time ranges.
-            let max_age = self.max_idle_age.as_secs();
-            let now = self.now.now();
-            self.vals.retain(|_, n| {
-                let age = now - n.last_access();
-                age.as_secs() <= max_age
-            });
-
-            if self.vals.len() == self.capacity {
-                return Err(CapacityExhausted {
-                    capacity: self.capacity,
-                });
+            match self.expirations.poll() {
+                // The cache is at capacity but we are able to remove a value.
+                Ok(Async::Ready(Some(entry))) => {
+                    self.vals.remove(entry.get_ref());
+                }
+
+                // `Ready(None)` can only be returned when expirations is
+                // empty. We know `expirations` is not empty because `vals` is
+                // not empty.
+                Ok(Async::Ready(None)) => unreachable!(),
```

Comment:
Extremely nitpicky: it might be nice to add a textual description to the `unreachable!()` macro, rather than explaining in a comment? That way, if this ever _is_ hit, the panic message will help explain the bug. Take it or leave it.

Reply:
No not nitpicky all all! I intended to add a better description here than just keeping the explanation as a comment.

Something this has me thinking about though is we can make a `Cache` with `capacity = 0`. If that was the case, we could hit this branch because `vals.len() == capacity` and `expirations.poll()` always return `Async::Ready`.

I'm not sure if we want to prevent a cache capacity of zero, so maybe we should handle this?

Reply:
Ahh, hm. I suppose we should probably put a
```rust
if capacity == 0 {
    return Err(CapacityExhausted {
        capacity: self.capacity,
    });                   
}
```

at the top of this function?